packages/matching-demo/src/algorithm/spreadMatcher.ts

```diff
@@ -0,0 +1,302 @@
+        const matches = [] as Match<T, P>[][];
+
+        let i = 0;
+        while ((i += 1)) {
+            if (i > 10000) {
+                throw new Error(`
+                  Matching executed more than 10000 - it may mean, that something broke
+                  data: ${JSON.stringify(data)}
+                  matches: not displayed, because of potential length
+                `);
+            }
+
+            const allGroupsHaveVolume = [
+                data.entityGroups[0].some((v) => v.volume.gt(0)),
+                data.entityGroups[1].some((v) => v.volume.gt(0))
+            ].every(Boolean);
+
+            if (!allGroupsHaveVolume) {
+                break;
+            }
+
+            const roundInput = getRoundInput(data);
+
+            // No routes can be computed
+            // if there is some consumption/generation left,
+            // but it won't be used because of missing priorities
+            if (roundInput.routes.length === 0) {
+                break;
+            }
+
+            const roundMatches = matchRound(roundInput);
+
+            roundMatches.forEach((match) => {
+                const entity1 = data.entityGroups[0].find((e) => e === match.entities[0])!;
+                const entity2 = data.entityGroups[1].find((e) => e === match.entities[1])!;
+
+                entity1.volume = entity1.volume.sub(match.volume);
+                entity2.volume = entity2.volume.sub(match.volume);
+            });
+
+            matches.push(roundMatches);
+        }
+
+        const roundMatches = matches.map((matches) =>
+            matches.map((match) => ({
+                entities: match.entities.map((entity) => ({
+                    ...entity
+                })),
+                volume: BigNumber.from(match.volume)
+            }))
+        ) as Match<T, P>[][];
+
+        return {
+            matches: sumMatches(matches.flat()).map(omitMatchVolumes) as Match<T, P>[],
+            roundMatches,
+            leftoverEntities: data.entityGroups
+        };
+    };
+
+    const omitMatchVolumes = <T extends Entity, P extends Entity>(match: Match<T, P>) => {
+        return {
+            ...match,
+            entities: match.entities.map((e) => omit(e, 'volume'))
+        };
+    };
+
+    /**
+     * Multiple matches for the same entities can happen,
+     * and we want sum of these.
+     */
+    const sumMatches = <T extends Entity, P extends Entity>(
+        matches: Match<T, P>[]
+    ): Match<T, P>[] => {
+        const areMatchesSame = (match1: Match<T, P>, match2: Match<T, P>) => {
+            return (
+                match1.entities[0] === match2.entities[0] &&
+                match1.entities[1] === match2.entities[1]
+            );
+        };
+
+        return matches.reduce((sum, match) => {
+            const existingMatch = sum.find((s) => areMatchesSame(s, match));
+
+            if (existingMatch) {
+                existingMatch.volume = existingMatch.volume.add(match.volume);
```

Comment:
What is the purpose of this assignment? 
As far as I understand it is not mutating the `existingMatch`? 
and you simply return the `sum`

Reply:
Don't dive into this algorithm please ;) This function is not clean TBH, I must agree (it does reduce + sideeffect), I may look into it when I will be touching algo soon